internal/controller/clientprofile_controller.go

```diff
@@ -0,0 +1,337 @@
+}
+
+const (
+	cleanupFinalizer = "csi.ceph.com/cleanup"
+)
+
+var configMapUpdateLock = sync.Mutex{}
+
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=clientprofile,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=clientprofile/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=clientprofile/finalizers,verbs=update
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=cephconnection,verbs=get;list;watch;update
+//+kubebuilder:rbac:groups="",resources=configmap,verbs=get;list;watch;create;update;patch;delete
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ClientProfileReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Filter update events based on metadata.generation changes, will filter events
+	// for non-spec changes on most resource types.
+	genChangedPredicate := predicate.GenerationChangedPredicate{}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&csiv1a1.ClientProfile{}).
+		Owns(
+			&csiv1a1.CephConnection{},
+			builder.MatchEveryOwner,
+			builder.WithPredicates(genChangedPredicate),
+		).
+		Owns(
+			&corev1.ConfigMap{},
+			builder.MatchEveryOwner,
+			builder.WithPredicates(genChangedPredicate),
+		).
+		Complete(r)
+}
+
+func (r *ClientProfileReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx)
+	log.Info("Starting reconcile iteration for ClientProfile", "req", req)
+
+	reconcileHandler := ClientProfileReconcile{}
+	reconcileHandler.ClientProfileReconciler = *r
+	reconcileHandler.ctx = ctx
+	reconcileHandler.log = log
+	reconcileHandler.clientProfile.Name = req.Name
+	reconcileHandler.clientProfile.Namespace = req.Namespace
+
+	err := reconcileHandler.reconcile()
+	if err != nil {
+		log.Error(err, "ClientProfile reconciliation failed")
+	} else {
+		log.Info("ClientProfile reconciliation completed successfully")
+	}
+	return ctrl.Result{}, err
+}
+
+func (r *ClientProfileReconcile) reconcile() error {
+	if err := r.loadAndValidate(); err != nil {
+		return err
+	}
+
+	// Ensure a finalizer on the ClientProfile to allow proper clean up
+	if ctrlutil.AddFinalizer(&r.clientProfile, cleanupFinalizer) {
+		if err := r.Update(r.ctx, &r.clientProfile); err != nil {
+			r.log.Error(err, "Failed to add a cleanup finalizer on ClientProfile")
+			return err
+		}
+	}
+
+	if err := r.reconcileCephConnection(); err != nil {
+		return err
+	}
+	if err := r.reconcileCephCsiClusterInfo(); err != nil {
+		return err
+	}
+
+	if r.cleanUp {
+		ctrlutil.RemoveFinalizer(&r.clientProfile, cleanupFinalizer)
+		if err := r.Update(r.ctx, &r.clientProfile); err != nil {
+			r.log.Error(err, "Failed to add a cleanup finalizer on config resource")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *ClientProfileReconcile) loadAndValidate() error {
+	// Load the ClientProfile
+	if err := r.Get(r.ctx, client.ObjectKeyFromObject(&r.clientProfile), &r.clientProfile); err != nil {
```

Comment:
I see a pattern of loading operands across all controllers, however an event is always being assumed as the resource will exist but for deletion it may not and this could potentially result in not ending reconciles, isn't it?

Reply:
As discussed https://github.com/ceph/ceph-csi-operator/issues/78 opened a tracker for it. resolving it